frontend/catalyst/debug.py

```diff
@@ -24,11 +26,34 @@
 from catalyst.compiled_functions import CompiledFunction
 from catalyst.compiler import Compiler
 from catalyst.jax_primitives import print_p
+from catalyst.pennylane_extensions import callback
 from catalyst.tracing.contexts import EvaluationContext
 from catalyst.tracing.type_signatures import filter_static_args, promote_arguments
 from catalyst.utils.filesystem import WorkspaceManager
 
 
+def debug_print(fmt: str, *args, **kwargs) -> None:
+    """
+    This function has been modified from its original form in the JAX project at
+    https://github.com/google/jax/blob/5eeebf2c1829bf3c66c947b6e12464a779434e29/jax/_src/debugging.py#L269
+    version released under the Apache License, Version 2.0, with the following copyright notice:
+
+    Copyright 2022 The Jax Authors.
+    """
+    callback(partial(_format_print_callback, fmt))(*args, **kwargs)
```

Comment:
My suggestion here is that, rather than a new function, we simply upstream this implementation into the existing `catalyst.debug.print`.

This might also solve a problem we currently have where `catalyst.debug.print` does not work in Jupyter notebooks.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/b7ea323a46426dfe11390823172d2cdfbb46458e

Reply:
Also, I would like to suggest having two functions:

1. `print` similar to python's print
2. `printf` similar to what is implemented here.

Thoughts?

Reply:
@erick-xanadu what is your motivation for the two functions?

Reply:
To keep the previous behaviour. They have different semantics. One is equivalent to the print statement. The other one is equivalent to: 

```python
def debug.print(fmt: str, *args: PyTree[Array], **kwargs: PyTree[Array]) -> None:
  print(fmt.format(*args, **kwargs))
```

Reply:
yep! I guess the way I see it is from a user POV, it feels a bit like the same function 🤔 that is, `fmt` could simply be a string with no `{}`, and `*args` could be empty, and it acts like a print statement?